File: java/src/main/java/org/sammancoaching/NodeSysInfoPrinter.java
package org.sammancoaching;

import java.util.Map;

public class NodeSysInfoPrinter {
    public static String print(NodeSysInfo info) {
        StringBuilder result = new StringBuilder("NodeSysInfo:\n");
        result.append("  Software Versions:\n");
        for (int i = 0; i <= NodeSysInfo.INDEX_CPU_MAX; i++) {
            CNodeSystemVersion swversion = info.get(i);
            if (swversion.isValid()) {
                result.append("    CPU ")
                      .append(i)
                      .append(": ")
                      .append(swversion.toString())
                      .append("\n");
            }
        }
        Map<String, String> addInfos = info.getAdditionalInfos();
        if (!addInfos.isEmpty()) {
            result.append("  Additional Infos:\n");
            for (Map.Entry<String, String> entry : addInfos.entrySet()) {
                result.append("    ")
                      .append(entry.getKey())
                      .append(": ")
                      .append(entry.getValue())
                      .append("\n");
            }
        }
        return result.toString();
    }
}

File: cpp/src/NodeSysInfoPrinter.cpp
#include "NodeSysInfoPrinter.hpp"
#include "NodeSysInfo.hpp"

std::string NodeSysInfoPrinter::print(const NodeSysInfo& info) {
    std::string result = "NodeSysInfo:\n";
    result += "  Software Versions:\n";
    for (size_t i = 0; i <= NodeSysInfo::index_cpu_max; ++i) {
        const auto& swversion = info.get(i);
        if (swversion.isValid()) {
            result += "    CPU " + std::to_string(i) + ": " + swversion.to_string() + "\n";
        }
    }
    if (!info.get_AddInfos().empty()) {
        result += "  Additional Infos:\n";
        for (auto const& [key, value] : info.get_AddInfos()) {
            result += "    " + key + ": " + value + "\n";
        }
    }
    return result;
}
